### ISSUE TITLE
Incorrect column number displayed for loooong lines

### ISSUE THREAD

**Issue by critiqjo** [martanne/vis#223]
Steps to reproduce:

1. Execute this in shell:
   ```sh
   $ for i in {0..99999}; do printf " $i"; done > vis-bug
   $ vis vis-bug
   ```

2. Press `$`
   The column number shown is 1, when it should be 588890 (or ..91).

3. Press `h`
   The column number shown is relative to the first character displayed on screen, not relative to the first character of the line.

Follow-up bug report: Incorrect behavior of `j` and `k` keys:

1. Open the same generated file
2. Press `$` followed by repeated `k`.

`k` behaves somewhat like pressing `h`, but slower. If I press and hold `k`, then it'll freeze for some time (due to 100% CPU usage -- try lowering your max-clock speed if you have a high-end CPU).

Commit: 02c6df7

-- bug report end; rants begin --

[...] _truncated a very long rant about stars!_

Performance: I tried the same with 10^6 and opened it with Vim: `vim -u NONE vis-bug`. Vim gets real slow when you do `$bbb`... But Vis handles it like a breeze. :+1: x10

_EDIT:_ Rants shortened, bug report lengthened

**Comment by arames** [comment]
The reason for the increasing stars: it is featured on hacker news.

**Comment by critiqjo** [comment]
> The reason for the increasing stars: it is featured on hacker news.

Phew! You saved me from a breakdown! :smile:

**Comment by critiqjo** [comment]
I have updated the original post to include a related bug report...

**Comment by joshaw** [comment]
I've come across another issue that I think might be related to this, so I'll add it here.

For a file with long lines that wrap, place a wrapped line at the bottom of the screen so that the last lines of the screen are:
- screen line with start of file line (with cursor on first character)
- screen line with wrapped portion of previous line
- statusline

Now, `j` moves the cursor to the wrapped line (as `gj` should do) rather than the next file line and a further k will move to the previous file line.

If the cursor is after the second char, then `j` causes the screen to jump down so the cursor is now at the top of the screen.

**Comment by martanne** [comment]
@critiqjo the wrong/relative column number was there for performance reasons. With the current code, displaying the correct value is expensive for very long lines.

However I agree that displaying wrong/misleading data is not a good idea. Therefore the column number is now hidden once a very long line was encountered. The same was already done for line numbers in very large files. The [used limits](https://github.com/martanne/vis/blob/c5f1b4e0f55fecf6c13f9a6fcdf6f0d44592171f/ui.h#L7) are currently hard coded and completely arbitrary.

With the current implementation `j` and `k` are quite inefficient when used on very long lines. This is illustrated by your test case (1 single line file) where they scan the whole file only to find out that there exist no other lines. It is basically an expensive NOP. For now use `gj` and `gk` instead.

Excellent large file support is a goal of vis and once I have a bit more time I will try to improve it further.

**Comment by martanne** [comment]
@joshaw your issue wasn't really related, but should now be "fixed" by efe5f155bf526be4448ed8b4a0cda9072e107a79. I'm not yet sure whether I like the new behavior though.

Anyway closing this, because the initial issue (wrong column number) has been addressed.
